File: src/football_cv/utils/geometry.py
"""
Geometric and bounding-box utility functions.
"""

from collections.abc import Sequence

Point2D = Sequence[float] | tuple[float, float]
BBox = Sequence[float] | tuple[float, float, float, float]


def get_center_of_bbox(bbox: BBox) -> tuple[int, int]:
    """
    Calculate the center coordinate (x, y) of a bounding box.

    Args:
        bbox: [x1, y1, x2, y2]

    Returns:
        (center_x, center_y) as integers.
    """
    x1, y1, x2, y2 = bbox[:4]
    center_x = (x1 + x2) / 2.0
    center_y = (y1 + y2) / 2.0
    return int(center_x), int(center_y)


def get_bbox_width(bbox: BBox) -> float:
    """Calculate the horizontal width of a bounding box."""
    return float(bbox[2] - bbox[0])


def get_bbox_height(bbox: BBox) -> float:
    """Calculate the vertical height of a bounding box."""
    return float(bbox[3] - bbox[1])


def get_foot_position(bbox: BBox) -> tuple[int, int]:
    """
    Calculate the bottom-center coordinate representing player foot contact.

    Args:
        bbox: [x1, y1, x2, y2]

    Returns:
        (foot_x, foot_y) as integers.
    """
    x1, _, x2, y2 = bbox[:4]
    foot_x = (x1 + x2) / 2.0
    foot_y = y2
    return int(foot_x), int(foot_y)


def measure_distance(p1: Point2D, p2: Point2D) -> float:
    """Calculate Euclidean distance between two 2D points."""
    return float(((p1[0] - p2[0]) ** 2 + (p1[1] - p2[1]) ** 2) ** 0.5)


def measure_xy_distance(p1: Point2D, p2: Point2D) -> tuple[float, float]:
    """Calculate signed (dx, dy) displacement from p2 to p1."""
    return float(p1[0] - p2[0]), float(p1[1] - p2[1])


def extract_centers(bboxes: Sequence[BBox]) -> list[tuple[int, int]]:
    """Extract center points for a collection of bounding boxes."""
    return [get_center_of_bbox(b) for b in bboxes]


def extract_foot_positions(bboxes: Sequence[BBox]) -> list[tuple[int, int]]:
    """Extract foot contact points for a collection of bounding boxes."""
    return [get_foot_position(b) for b in bboxes]


def filter_valid_bboxes(bboxes: Sequence[BBox]) -> list[BBox]:
    """Filter out bounding boxes with non-finite coordinates or negative dimensions."""
    import math

    valid: list[BBox] = []
    for b in bboxes:
        if len(b) < 4:
            continue
        x1, y1, x2, y2 = b[:4]
        if not (
            math.isfinite(x1)
            and math.isfinite(y1)
            and math.isfinite(x2)
            and math.isfinite(y2)
        ):
            continue
        if x2 >= x1 and y2 >= y1:
            valid.append(b)
    return valid


def find_nearest_point(
    target: Point2D,
    candidates: Sequence[Point2D],
    max_distance: float = float("inf"),
) -> int | None:
    """Find the index of the nearest candidate point within max_distance."""
    best_idx: int | None = None
    min_dist = max_distance

    for idx, cand in enumerate(candidates):
        dist = measure_distance(target, cand)
        if dist < min_dist:
            min_dist = dist
            best_idx = idx

    return best_idx

File: cpp/src/geometry.cpp
#include "football_cv/geometry.hpp"

#include <algorithm>
#include <cmath>

namespace football_cv {

Point2D get_center_of_bbox(const BoundingBox& bbox) noexcept {
    return bbox.center();
}

Point2D get_foot_position(const BoundingBox& bbox) noexcept {
    return bbox.foot_position();
}

double get_bbox_width(const BoundingBox& bbox) noexcept {
    return bbox.width();
}

double get_bbox_height(const BoundingBox& bbox) noexcept {
    return bbox.height();
}

double measure_distance(const Point2D& p1, const Point2D& p2) noexcept {
    return p1.distance_to(p2);
}

std::pair<double, double> measure_xy_distance(const Point2D& p1, const Point2D& p2) noexcept {
    // Matches Python measure_xy_distance(p1, p2) -> (p1[0] - p2[0], p1[1] - p2[1])
    return {p1.x - p2.x, p1.y - p2.y};
}

std::vector<Point2D> extract_centers(const std::vector<BoundingBox>& bboxes) {
    std::vector<Point2D> centers;
    centers.reserve(bboxes.size());
    for (const auto& box : bboxes) {
        centers.push_back(box.center());
    }
    return centers;
}

std::vector<Point2D> extract_foot_positions(const std::vector<BoundingBox>& bboxes) {
    std::vector<Point2D> feet;
    feet.reserve(bboxes.size());
    for (const auto& box : bboxes) {
        feet.push_back(box.foot_position());
    }
    return feet;
}

std::vector<BoundingBox> filter_valid_bboxes(const std::vector<BoundingBox>& bboxes) {
    std::vector<BoundingBox> valid;
    valid.reserve(bboxes.size());
    for (const auto& box : bboxes) {
        if (box.is_valid()) {
            valid.push_back(box);
        }
    }
    return valid;
}

std::optional<size_t> find_nearest_point(
    const Point2D& target,
    const std::vector<Point2D>& candidates,
    double max_distance
) {
    if (!target.is_finite() || candidates.empty() || max_distance < 0.0) {
        return std::nullopt;
    }

    std::optional<size_t> best_idx = std::nullopt;
    double min_dist = max_distance;

    for (size_t i = 0; i < candidates.size(); ++i) {
        const auto& pt = candidates[i];
        if (!pt.is_finite()) {
            continue;
        }
        const double d = target.distance_to(pt);
        if (d <= min_dist) {
            min_dist = d;
            best_idx = i;
        }
    }

    return best_idx;
}

} // namespace football_cv
